### PR TITLE
Optical bounds

### PR DESCRIPTION
**Pull request by behdad**
Fixes https://github.com/harfbuzz/harfbuzz/issues/3458

Untested.

### CODE REVIEW

File: src/hb-ot-layout.cc [hunk 1]
```diff
@@ -1703,6 +1703,63 @@ hb_ot_layout_get_size_params (hb_face_t       *face,
...
+  dispatch (const T &obj, Ts&&... ds) HB_AUTO_RETURN
+  ( _dispatch (obj, hb_prioritize, std::forward<Ts> (ds)...) )
+};
+
+hb_position_t
+hb_ot_layout_get_optical_bound (hb_font_t      *font,
+				unsigned        lookup_index,
+				hb_direction_t  direction,
+				hb_codepoint_t  glyph)
+{
```
Comment (matthiasclasen) [inline src/hb-ot-layout.cc:1731]: My immediate questions, trying to use this api:
- How do I interpret the value that is returned ? Is it 'how far the glyph should extend into the margin' or something else?
- Where do I get this lookup_index you want?
- Can we add some boolean for 'has optical bound information' ? Or maybe that would be answered by trying to find that lookup_index?

Comment (behdad) [reply src/hb-ot-layout.cc:1731]: > * How do I interpret the value that is returned ? Is it 'how far the glyph should extend into the margin' or something else?

Right. Though, it's *negative* for extending into the margin. That's just how the numbers in the font were and I kept them that way. So basically that's how much you add to the x_offset and advance_width, for the left side, and you add to advance_width, for the right side.

> * Where do I get this lookup_index you want?

Ultimately from `hb_ot_layout_feature_get_lookups`. You need to get feature index first.

> * Can we add some boolean for 'has optical bound information' ? Or maybe that would be answered by trying to find that lookup_index?

Yeah, if there was no lookup_index you know...  Or even earlier, there would be no feature_index.

Comment (matthiasclasen) [reply src/hb-ot-layout.cc:1731]: I've got lucky with hb_ot_layout_collect_lookups

File: src/hb-ot-layout.cc [hunk 2]
```diff
@@ -1703,6 +1703,63 @@ hb_ot_layout_get_size_params (hb_face_t       *face,
...
+  /* This is ugly... */
+  hb_buffer_t buffer;
+  buffer.props.direction = direction;
+  OT::hb_ot_apply_context_t c_apply (1, font, &buffer);
+  const OT::PosLookup &lookup = font->face->table.GPOS->table->get_lookup (lookup_index);
+  hb_glyph_position_t pos = {0};
+  hb_position_single_dispatch_t c;
+  lookup.dispatch (&c, &c_apply, glyph, pos);
+  hb_position_t ret = 0;
+  switch (direction)
```
Comment (matthiasclasen) [inline src/hb-ot-layout.cc:1737]: Do you think this is a misuse of direction? What we really want here is a side (ie left/right/top/bottom), not a direction. 
The right bounds will be used in ltr context too...

Comment (behdad) [reply src/hb-ot-layout.cc:1737]: I know... But another enum with exactly four values sounds a bit overkill...

Comment (behdad) [reply src/hb-ot-layout.cc:1737]: Think of it as "left-to-right edge of the glyph".

Comment (matthiasclasen) [reply src/hb-ot-layout.cc:1737]: but then, there are no features for top/bottom bounds anyway

Comment (behdad) [reply src/hb-ot-layout.cc:1737]: In theory there could; just needs a feature tag (which could be private) to be named.

**Comment by matthiasclasen** [comment]
I can confirm that this api works, tested with this pango code: https://gitlab.gnome.org/GNOME/pango/-/commit/3fedb89e7b2301d4de0f3885bd5e2ed9cfe38105
and Linux Libertine fonts.

File: src/hb-ot-layout.cc (hunk 1 above)
Comment (matthiasclasen) [reply src/hb-ot-layout.cc:1731]: We should probably turn this discussion into a doc comment

**Comment by behdad** [comment]
Thanks. I'll add docs, double-check, and merge.

File: src/hb-ot-layout.cc (hunk 1 above)
Comment (matthiasclasen) [reply src/hb-ot-layout.cc:1731]: Theoretical question: what would I do if ltbd returned more than one lookup ? try all until I find the glyph ?

Comment (behdad) [reply src/hb-ot-layout.cc:1731]: > Theoretical question: what would I do if ltbd returned more than one lookup ? try all until I find the glyph ?

Correct.

Comment (behdad) [reply src/hb-ot-layout.cc:1731]: > Theoretical question: what would I do if ltbd returned more than one lookup ? try all until I find the glyph ?

The loop you have ignores current script/lang. You might want to adjust it but is more work of course. You might as well keep it and loop over the lookups.

File: src/hb-ot-layout.cc [hunk 3]
```diff
@@ -1703,6 +1703,59 @@ hb_ot_layout_get_size_params (hb_face_t       *face,
...
+  template <typename T, typename ...Ts> auto
+  dispatch (const T &obj, Ts&&... ds) HB_AUTO_RETURN
+  ( _dispatch (obj, hb_prioritize, std::forward<Ts> (ds)...) )
+};
+
+hb_position_t
+hb_ot_layout_get_optical_bound (hb_font_t      *font,
+				unsigned        lookup_index,
+				hb_direction_t  direction,
+				hb_codepoint_t  glyph)
```
Comment (khaledhosny) [inline src/hb-ot-layout.cc:1730]: Applications like LuaTeX want all the optical bound values in advance (it has a mode where line breaking takes it into account). I guess one can use `hb_ot_layout_lookup_collect_glyphs()` to get all affected glyphs and iterate on them one by one, so would it be faster if the API took an array of glyphs (or to have a different API that return optical bounds for all glyphs in the lookup?)

Comment (behdad) [reply src/hb-ot-layout.cc:1730]: Humm. That complicates things. Yeah I'm leaning towards suggesting collect_glyphs first.

Comment (behdad) [reply src/hb-ot-layout.cc]: > Applications like LuaTeX want all the optical bound values in advance

We can add a version that works on an array later if that fetching one by one seems to be too slow.

**Pull request merged**